robot/web/stream/state/state.go

```diff
@@ -171,296 +145,214 @@ func (mt msgType) String() string {
-			return fmt.Errorf("unexpected stream %s source %s", ss.Stream.Name(), ss.streamSource)
-		}
+func (state *StreamState) tick() {
+	switch {
+	case state.activeClients < 0:
+		state.logger.Error("activeClients is less than 0")
+	case state.activeClients == 0:
+		// stop stream if there are no active clients
+		// noop if there is no stream source
+		state.stopInputStream()
+	case state.streamSource == streamSourceUnknown: // && state.activeClients > 0
 		// this is the first subscription, attempt passthrough
-		ss.logger.CDebugw(ctx, "attempting to subscribe to rtp_passthrough", "name", ss.Stream.Name())
-		err := ss.streamH264Passthrough(ctx)
+		state.logger.Info("attempting to subscribe to rtp_passthrough")
+		err := state.streamH264Passthrough()
 		if err != nil {
-			ss.logger.CDebugw(ctx, "rtp_passthrough not possible, falling back to GoStream", "err", err.Error(), "name", ss.Stream.Name())
+			state.logger.Warnw("tick: rtp_passthrough not possible, falling back to GoStream", "err", err)
 			// if passthrough failed, fall back to gostream based approach
-			ss.Stream.Start()
-			ss.streamSource = streamSourceGoStream
+			state.Stream.Start()
+			state.streamSource = streamSourceGoStream
 		}
-		ss.activePeers++
-		return nil
-	}
+	case state.streamSource == streamSourcePassthrough && state.streamSourceSub.Terminated.Err() != nil:
+		// restart stream if there we were using passthrough but the sub is terminated
+		state.logger.Info("previous subscription terminated attempting to subscribe to rtp_passthrough")
 
-	switch ss.streamSource {
-	case streamSourcePassthrough:
-		ss.logger.CDebugw(ctx, "continuing using rtp_passthrough", "name", ss.Stream.Name())
-		// noop as we are already subscribed
-	case streamSourceGoStream:
-		ss.logger.CDebugw(ctx, "currently using gostream, trying upgrade to rtp_passthrough", "name", ss.Stream.Name())
-		// attempt to cut over to passthrough
-		err := ss.streamH264Passthrough(ctx)
+		err := state.streamH264Passthrough()
 		if err != nil {
-			ss.logger.CDebugw(ctx, "rtp_passthrough not possible, continuing with gostream", "err", err.Error(), "name", ss.Stream.Name())
-		}
-	case streamSourceUnknown:
-		fallthrough
-	default:
-		err := fmt.Errorf("%s streamSource in unexpected state %s", ss.Stream.Name(), ss.streamSource)
-		ss.logger.Error(err.Error())
-		return err
-	}
-	ss.activePeers++
-	return nil
-}
-
-func (ss *StreamState) dec(ctx context.Context) error {
-	ss.logger.Debugf("decrement START %s activePeers: %d", ss.Stream.Name(), ss.activePeers)
-	defer func() { ss.logger.Debugf("decrement END %s activePeers: %d", ss.Stream.Name(), ss.activePeers) }()
-
-	var err error
-	defer func() {
-		if err != nil {
-			ss.logger.Errorf("decrement %s hit error: %s", ss.Stream.Name(), err.Error())
-			return
-		}
-		ss.activePeers--
-		if ss.activePeers <= 0 {
-			ss.activePeers = 0
+			state.logger.Warn("rtp_passthrough not possible, falling back to GoStream", "err", err)
```

Comment:
```suggestion
			state.logger.Info("rtp_passthrough not possible, falling back to GoStream", "err", err)
```

I don't think we want to emit warn logs for cameras like webcam, how about we just emit this at info level?